tests/Tests/ORM/Mapping/NamingStrategy/CustomPascalNamingStrategy.php

```diff
@@ -0,0 +1,101 @@
+ * regarding fully custom naming strategies.
+ */
+class CustomPascalNamingStrategy implements NamingStrategy
+{
+    /**
+     * Returns a table name for an entity class.
+     *
+     * @param string $className The fully-qualified class name
+     * @return string A table name
+     */
+    public function classToTableName(string $className): string
+    {
+        if (false !== strpos($className, '\\')) {
+            return substr($className, strrpos($className, '\\') + 1);
+        }
+
+        return $className;
+    }
+
+    /**
+     * Returns a column name for a property.
+     *
+     * @param string $propertyName A property name
+     * @param string|null $className    The fully-qualified class name
+     *
+     * @return string A column name
+     */
+    public function propertyToColumnName(string $propertyName, ?string $className = null): string
+    {
+        if (null !== $className && strtolower($propertyName) == strtolower($this->classToTableName($className)) . 'id') {
+            return 'Id';
+        }
+
+        return ucfirst($propertyName);
+    }
+
+    /**
+     * Returns a column name for an embedded property.
+     */
+    public function embeddedFieldToColumnName(string $propertyName, string $embeddedColumnName, ?string $className = null, $embeddedClassName = null): string
+    {
+        throw new \LogicException(sprintf('Method %s is not implemented', __METHOD__));
+    }
+
+    /**
+     * Returns the default reference column name.
+     *
+     * @return string A column name
+     */
+    public function referenceColumnName(): string
+    {
+        return 'Id';
+    }
+
+    /**
+     * Returns a join column name for a property.
+     *
+     * @return string A join column name
+     */
+    public function joinColumnName(string $propertyName, string $className): string
+    {
+        return ucfirst($propertyName) . $this->referenceColumnName();
+    }
+
+    /**
+     * Returns a join table name.
+     *
+     * @param string $sourceEntity The source entity
+     * @param string $targetEntity The target entity
+     * @param string|null $propertyName A property name
+     *
+     * @return string A join table name
+     */
+    public function joinTableName(string $sourceEntity, string $targetEntity, ?string $propertyName = null): string
+    {
+        return $this->classToTableName($sourceEntity) . $this->classToTableName($targetEntity);
+    }
+
+    /**
+     * Returns the foreign key column name for the given parameters.
+     *
+     * @param string      $entityName           An entity
+     * @param string|null $referencedColumnName A property
+     *
+     * @return string A join column name
+     */
+    public function joinKeyColumnName(string $entityName, ?string $referencedColumnName = null): string
+    {
+        return $this->classToTableName($entityName) . ($referencedColumnName ?: $this->referenceColumnName());
+    }
+}
```

Comment:
Please configure your IDE to always add a newline at end of file, this will make the red pictogram disappear.

Reply:
My IDE doesn't do that on purpose as I don't like it changing code while working - PHP CS Fixer should do that  ;)